In Rocq output, implicit arguments to `typeName.t : Type` are not cleared for non-struct generics

It appears that Goose might not properly translate the usage of type parameters as type arguments (at least in structs).

Given the following Go module, structured as follows:

`failing_example/example.go`:
```go
package example

type items[T any] []T

type node[T any] struct {
	items    items[T]
}
```
`failing_example/go.mod`:

```go
module example

go 1.18
```

Next, I run the command `goose -out perennial/new/code -dir failing_example`. Then,
when running `make new/code/example.vo` in my copy of Perennial, I get the following error:
```
ROCQ DEP .rocqdeps.d
ROCQ COMPILE new/code/example.v
File "./new/code/example.v", line 48, characters 12-29:
Error: Illegal application (Non-functional construction): 
The expression "items.t" of type "Type" cannot be applied to the term
 "T" : "Type"

make: *** [Makefile:66: new/code/example.vo] Error 1
```

Here's the text of the generated [example.v](https://gist.github.com/codyjrivera/46d083e7ff3849c84ad0c98f0727586b).

--- a/testdata/examples/append_log/append_log.gold.v
+++ b/testdata/examples/append_log/append_log.gold.v
@@ -212,7 +212,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Log.
 
 Definition Log'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [

--- a/testdata/examples/channel/channel.gold.v
+++ b/testdata/examples/channel/channel.gold.v
@@ -1957,7 +1957,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Cond.
 
 Definition Cond'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -2000,7 +1999,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End LockedStack.
 
 Definition LockedStack'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -2038,7 +2036,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End EliminationStack.
 
 Definition EliminationStack'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -2076,7 +2073,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End request.
 
 Definition request'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -2111,7 +2107,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Lock.
 
 Definition Lock'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -2152,7 +2147,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End stream.
 
 Definition stream'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -2189,7 +2183,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End streamold.
 
 Definition streamold'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [

--- a/testdata/examples/comments/comments.gold.v
+++ b/testdata/examples/comments/comments.gold.v
@@ -38,7 +38,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Foo.
 
 Definition Foo'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [

--- a/testdata/examples/interfacerecursion/interfacerecursion.gold.v
+++ b/testdata/examples/interfacerecursion/interfacerecursion.gold.v
@@ -54,6 +54,7 @@ Module A.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := interface.t.
+#[global] Arguments t : clear implicits.
 End def.
 End A.
 
@@ -68,6 +69,7 @@ Module B.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := interface.t.
+#[global] Arguments t : clear implicits.
 End def.
 End B.
 
@@ -89,7 +91,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End c.
 
 Definition c'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [

--- a/testdata/examples/logging2/logging2.gold.v
+++ b/testdata/examples/logging2/logging2.gold.v
@@ -391,7 +391,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Log.
 
 Definition Log'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -466,7 +465,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Txn.
 
 Definition Txn'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [

--- a/testdata/examples/mapliteral/mapliteral.gold.v
+++ b/testdata/examples/mapliteral/mapliteral.gold.v
@@ -59,7 +59,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Nested.
 
 Definition Nested'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [

--- a/testdata/examples/semantics/semantics.gold.v
+++ b/testdata/examples/semantics/semantics.gold.v
@@ -3323,7 +3323,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End unit.
 
 Definition unit'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3354,7 +3353,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Editor.
 
 Definition Editor'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3391,7 +3389,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Pair.
 
 Definition Pair'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3418,6 +3415,7 @@ Module Uint32.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := w32.
+#[global] Arguments t : clear implicits.
 End def.
 End Uint32.
 
@@ -3432,6 +3430,7 @@ Module geometryInterface.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := interface.t.
+#[global] Arguments t : clear implicits.
 End def.
 End geometryInterface.
 
@@ -3454,7 +3453,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End SquareStruct.
 
 Definition SquareStruct'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3490,7 +3488,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End NumStruct.
 
 Definition NumStruct'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3514,6 +3511,7 @@ Module shapeInterface.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := interface.t.
+#[global] Arguments t : clear implicits.
 End def.
 End shapeInterface.
 
@@ -3528,6 +3526,7 @@ Module polygonInterface.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := interface.t.
+#[global] Arguments t : clear implicits.
 End def.
 End polygonInterface.
 
@@ -3550,7 +3549,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End shapeStruct.
 
 Definition shapeStruct'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3585,7 +3583,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End polygonStruct.
 
 Definition polygonStruct'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3616,6 +3613,7 @@ Module dogInterface.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := interface.t.
+#[global] Arguments t : clear implicits.
 End def.
 End dogInterface.
 
@@ -3630,6 +3628,7 @@ Module catInterface.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := interface.t.
+#[global] Arguments t : clear implicits.
 End def.
 End catInterface.
 
@@ -3644,6 +3643,7 @@ Module Puppy.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := go_string.
+#[global] Arguments t : clear implicits.
 End def.
 End Puppy.
 
@@ -3662,6 +3662,7 @@ Module Kitten.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := go_string.
+#[global] Arguments t : clear implicits.
 End def.
 End Kitten.
 
@@ -3680,6 +3681,7 @@ Module printInterface.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := interface.t.
+#[global] Arguments t : clear implicits.
 End def.
 End printInterface.
 
@@ -3702,7 +3704,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End PaperStruct.
 
 Definition PaperStruct'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3728,6 +3729,7 @@ Module Flower.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := interface.t.
+#[global] Arguments t : clear implicits.
 End def.
 End Flower.
 
@@ -3742,6 +3744,7 @@ Module Flora.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := interface.t.
+#[global] Arguments t : clear implicits.
 End def.
 End Flora.
 
@@ -3763,7 +3766,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Lily.
 
 Definition Lily'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3796,7 +3798,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Rose.
 
 Definition Rose'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3829,7 +3830,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Daisy.
 
 Definition Daisy'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3863,7 +3863,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End LoopStruct.
 
 Definition LoopStruct'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3900,7 +3899,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End BoolTest.
 
 Definition BoolTest'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3942,7 +3940,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End ArrayEditor.
 
 Definition ArrayEditor'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3979,7 +3976,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Bar.
 
 Definition Bar'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -4015,7 +4011,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Foo.
 
 Definition Foo'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -4049,7 +4044,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End TwoInts.
 
 Definition TwoInts'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -4086,7 +4080,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End S.
 
 Definition S'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -4130,7 +4123,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End StructWrap.
 
 Definition StructWrap'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -4162,7 +4154,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End StructWithFunc.
 
 Definition StructWithFunc'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -4193,7 +4184,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End switchConcrete.
 
 Definition switchConcrete'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -4216,6 +4206,7 @@ Module switchInterface.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := interface.t.
+#[global] Arguments t : clear implicits.
 End def.
 End switchInterface.
 
@@ -4230,6 +4221,7 @@ Module DefinedStr.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := go_string.
+#[global] Arguments t : clear implicits.
 End def.
 End DefinedStr.
 
@@ -4253,7 +4245,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End List.
 
 Definition List'fds_unsealed (T : go.type) {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -4291,7 +4282,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Log.
 
 Definition Log'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [

--- a/testdata/examples/simpledb/simpledb.gold.v
+++ b/testdata/examples/simpledb/simpledb.gold.v
@@ -1055,7 +1055,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Table.
 
 Definition Table'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -1091,7 +1090,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Entry.
 
 Definition Entry'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -1127,7 +1125,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End lazyFileBuf.
 
 Definition lazyFileBuf'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -1163,7 +1160,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End bufFile.
 
 Definition bufFile'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -1201,7 +1197,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End tableWriter.
 
 Definition tableWriter'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -1248,7 +1243,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Database.
 
 Definition Database'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [

--- a/testdata/examples/unittest/generics/generics.go
+++ b/testdata/examples/unittest/generics/generics.go
@@ -97,3 +97,9 @@ func useAnyPointer() {
 	var x uint64
 	helpers.AnyPointer(&x)
 }
+
+type nonStructGeneric[T any] []T
+
+type useNonStructGeneric[T any] struct {
+	x nonStructGeneric[T]
+}

--- a/testdata/examples/unittest/generics/generics.gold.v
+++ b/testdata/examples/unittest/generics/generics.gold.v
@@ -28,6 +28,14 @@ Definition MultiParam {ext : ffi_syntax} {go_gctx : GoGlobalContext} (A : go.typ
 
 #[global] Opaque MultiParam.
 
+Definition nonStructGeneric {ext : ffi_syntax} {go_gctx : GoGlobalContext} (T : go.type) : go.type := go.Named "github.com/goose-lang/goose/testdata/examples/unittest/generics.nonStructGeneric"%go [T].
+
+#[global] Opaque nonStructGeneric.
+
+Definition useNonStructGeneric {ext : ffi_syntax} {go_gctx : GoGlobalContext} (T : go.type) : go.type := go.Named "github.com/goose-lang/goose/testdata/examples/unittest/generics.useNonStructGeneric"%go [T].
+
+#[global] Opaque useNonStructGeneric.
+
 Definition UnderlyingSlice {ext : ffi_syntax} {go_gctx : GoGlobalContext} : go_string := "github.com/goose-lang/goose/testdata/examples/unittest/generics.UnderlyingSlice"%go.
 
 Definition Clone {ext : ffi_syntax} {go_gctx : GoGlobalContext} : go_string := "github.com/goose-lang/goose/testdata/examples/unittest/generics.Clone"%go.
@@ -210,7 +218,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Box.
 
 Definition Box'fds_unsealed (T : go.type) {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -247,7 +254,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Container.
 
 Definition Container'fds_unsealed (T : go.type) {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -288,7 +294,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End UseContainer.
 
 Definition UseContainer'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -321,7 +326,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End OnlyIndirect.
 
 Definition OnlyIndirect'fds_unsealed (T : go.type) {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -357,7 +361,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End MultiParam.
 
 Definition MultiParam'fds_unsealed (A : go.type) (B : go.type) {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -380,6 +383,52 @@ Class MultiParam_Assumptions {ext : ffi_syntax} `{!GoGlobalContext} `{!GoLocalCo
   #[global] MultiParam_set_X A B A' B' (x : MultiParam.t A' B') y :: ⟦StructFieldSet (MultiParamⁱᵐᵖˡ A B) "X", (#x, #y)⟧ ⤳[under] #(x <|MultiParam.X' := y|>);
 }.
 
+Module nonStructGeneric.
+Section def.
+Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
+Definition t {T : Type} : Type := slice.t.
+#[global] Arguments t : clear implicits.
+End def.
+End nonStructGeneric.
+
+Definition nonStructGenericⁱᵐᵖˡ {ext : ffi_syntax} {go_gctx : GoGlobalContext} (T : go.type) : go.type := go.SliceType T.
+
+Class nonStructGeneric_Assumptions {ext : ffi_syntax} `{!GoGlobalContext} `{!GoLocalContext} `{!GoSemanticsFunctions} : Prop :=
+{
+  #[global] nonStructGeneric_underlying T :: (nonStructGeneric T) <u (nonStructGenericⁱᵐᵖˡ T);
+}.
+
+Module useNonStructGeneric.
+Section def.
+Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
+Record t {T : Type} :=
+mk {
+  x' : (generics.nonStructGeneric.t T);
+}.
+
+#[global] Instance zero_val `{!ZeroVal T} : ZeroVal t := {| zero_val := mk T (zero_val _)|}.
+#[global] Arguments mk : clear implicits.
+#[global] Arguments t : clear implicits.
+End def.
+End useNonStructGeneric.
+
+Definition useNonStructGeneric'fds_unsealed (T : go.type) {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
+  (go.FieldDecl "x"%go (nonStructGeneric T))
+].
+Program Definition useNonStructGeneric'fds (T : go.type) {ext : ffi_syntax} {go_gctx : GoGlobalContext} := sealed (useNonStructGeneric'fds_unsealed T).
+Global Instance equals_unfold_useNonStructGeneric T {ext : ffi_syntax} {go_gctx : GoGlobalContext} : useNonStructGeneric'fds T =→ useNonStructGeneric'fds_unsealed T.
+Proof. rewrite /useNonStructGeneric'fds seal_eq //. Qed.
+
+Definition useNonStructGenericⁱᵐᵖˡ {ext : ffi_syntax} {go_gctx : GoGlobalContext} (T : go.type) : go.type := go.StructType (useNonStructGeneric'fds T).
+
+Class useNonStructGeneric_Assumptions {ext : ffi_syntax} `{!GoGlobalContext} `{!GoLocalContext} `{!GoSemanticsFunctions} : Prop :=
+{
+  #[global] useNonStructGeneric_type_repr T T' `{!ZeroVal T'} `{!TypeRepr T T'} :: go.TypeReprUnderlying (useNonStructGenericⁱᵐᵖˡ T) (useNonStructGeneric.t T');
+  #[global] useNonStructGeneric_underlying T :: (useNonStructGeneric T) <u (useNonStructGenericⁱᵐᵖˡ T);
+  #[global] useNonStructGeneric_get_x T T' (x : useNonStructGeneric.t T') :: ⟦StructFieldGet (useNonStructGenericⁱᵐᵖˡ T) "x", #x⟧ ⤳[under] #x.(useNonStructGeneric.x');
+  #[global] useNonStructGeneric_set_x T T' (x : useNonStructGeneric.t T') y :: ⟦StructFieldSet (useNonStructGenericⁱᵐᵖˡ T) "x", (#x, #y)⟧ ⤳[under] #(x <|useNonStructGeneric.x' := y|>);
+}.
+
 Class Assumptions {ext : ffi_syntax} `{!GoGlobalContext} `{!GoLocalContext} `{!GoSemanticsFunctions} : Prop :=
 {
   #[global] Box_instance :: Box_Assumptions;
@@ -387,6 +436,8 @@ Class Assumptions {ext : ffi_syntax} `{!GoGlobalContext} `{!GoLocalContext} `{!G
   #[global] UseContainer_instance :: UseContainer_Assumptions;
   #[global] OnlyIndirect_instance :: OnlyIndirect_Assumptions;
   #[global] MultiParam_instance :: MultiParam_Assumptions;
+  #[global] nonStructGeneric_instance :: nonStructGeneric_Assumptions;
+  #[global] useNonStructGeneric_instance :: useNonStructGeneric_Assumptions;
   #[global] UnderlyingSlice_unfold T :: FuncUnfold UnderlyingSlice [T] (UnderlyingSliceⁱᵐᵖˡ T);
   #[global] Clone_unfold S E :: FuncUnfold Clone [S; E] (Cloneⁱᵐᵖˡ S E);
   #[global] BoxGet_unfold T :: FuncUnfold BoxGet [T] (BoxGetⁱᵐᵖˡ T);

--- a/testdata/examples/unittest/unittest.gold.v
+++ b/testdata/examples/unittest/unittest.gold.v
@@ -3193,6 +3193,7 @@ Module Foo.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := (array.t w64 10).
+#[global] Arguments t : clear implicits.
 End def.
 End Foo.
 
@@ -3214,7 +3215,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End importantStruct.
 
 Definition importantStruct'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3236,6 +3236,7 @@ Module stringWrapper.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := go_string.
+#[global] Arguments t : clear implicits.
 End def.
 End stringWrapper.
 
@@ -3250,6 +3251,7 @@ Module Uint32.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := w32.
+#[global] Arguments t : clear implicits.
 End def.
 End Uint32.
 
@@ -3264,6 +3266,7 @@ Module numWrapper.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := w64.
+#[global] Arguments t : clear implicits.
 End def.
 End numWrapper.
 
@@ -3287,7 +3290,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End withInterface.
 
 Definition withInterface'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3319,7 +3321,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End diskWrapper.
 
 Definition diskWrapper'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3351,7 +3352,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End embedA.
 
 Definition embedA'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3386,7 +3386,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End embedB.
 
 Definition embedB'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3422,7 +3421,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End embedC.
 
 Definition embedC'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3460,7 +3458,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End embedD.
 
 Definition embedD'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3490,6 +3487,7 @@ Module Enum1.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := w64.
+#[global] Arguments t : clear implicits.
 End def.
 End Enum1.
 
@@ -3504,6 +3502,7 @@ Module Enum2.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := w64.
+#[global] Arguments t : clear implicits.
 End def.
 End Enum2.
 
@@ -3518,6 +3517,7 @@ Module Fooer.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := interface.t.
+#[global] Arguments t : clear implicits.
 End def.
 End Fooer.
 
@@ -3540,7 +3540,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End concreteFooer.
 
 Definition concreteFooer'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3573,7 +3572,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End FooerUser.
 
 Definition FooerUser'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3597,6 +3595,7 @@ Module PointerInterface.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := interface.t.
+#[global] Arguments t : clear implicits.
 End def.
 End PointerInterface.
 
@@ -3618,7 +3617,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End concrete1.
 
 Definition concrete1'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3643,6 +3641,7 @@ Module my_u32.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := w32.
+#[global] Arguments t : clear implicits.
 End def.
 End my_u32.
 
@@ -3657,6 +3656,7 @@ Module also_u32.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := unittest.my_u32.t.
+#[global] Arguments t : clear implicits.
 End def.
 End also_u32.
 
@@ -3681,7 +3681,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End allTheLiterals.
 
 Definition allTheLiterals'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3719,7 +3718,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End hasCondVar.
 
 Definition hasCondVar'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3743,6 +3741,7 @@ Module IntWrapper.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := w64.
+#[global] Arguments t : clear implicits.
 End def.
 End IntWrapper.
 
@@ -3757,6 +3756,7 @@ Module MapWrapper.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := map.t.
+#[global] Arguments t : clear implicits.
 End def.
 End MapWrapper.
 
@@ -3780,7 +3780,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End mapElem.
 
 Definition mapElem'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3815,7 +3814,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End containsPointer.
 
 Definition containsPointer'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3847,7 +3845,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End wrapExternalStruct.
 
 Definition wrapExternalStruct'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3881,7 +3878,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End typing.
 
 Definition typing'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3914,7 +3910,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End composite.
 
 Definition composite'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3948,7 +3943,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End R.
 
 Definition R'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -3979,7 +3973,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Other.
 
 Definition Other'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -4013,7 +4006,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End RecursiveEmbedded.
 
 Definition RecursiveEmbedded'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -4046,7 +4038,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Block.
 
 Definition Block'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -4070,6 +4061,7 @@ Module SliceAlias.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := slice.t.
+#[global] Arguments t : clear implicits.
 End def.
 End SliceAlias.
 
@@ -4092,7 +4084,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End thing.
 
 Definition thing'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -4124,7 +4115,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End sliceOfThings.
 
 Definition sliceOfThings'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -4159,7 +4149,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Point.
 
 Definition Point'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -4201,7 +4190,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End TwoInts.
 
 Definition TwoInts'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -4238,7 +4226,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End S.
 
 Definition S'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -4283,7 +4270,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End B.
 
 Definition B'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -4314,7 +4300,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End A.
 
 Definition A'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [
@@ -4336,6 +4321,7 @@ Module Timestamp.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := w64.
+#[global] Arguments t : clear implicits.
 End def.
 End Timestamp.
 
@@ -4350,6 +4336,7 @@ Module UseTypeAbbrev.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := w64.
+#[global] Arguments t : clear implicits.
 End def.
 End UseTypeAbbrev.
 
@@ -4364,6 +4351,7 @@ Module UseNamedType.
 Section def.
 Context {ext : ffi_syntax} {go_gctx : GoGlobalContext}.
 Definition t : Type := unittest.Timestamp.t.
+#[global] Arguments t : clear implicits.
 End def.
 End UseNamedType.
 

--- a/testdata/examples/wal/wal.gold.v
+++ b/testdata/examples/wal/wal.gold.v
@@ -380,7 +380,6 @@ mk {
 #[global] Arguments mk : clear implicits.
 #[global] Arguments t : clear implicits.
 End def.
-
 End Log.
 
 Definition Log'fds_unsealed {ext : ffi_syntax} {go_gctx : GoGlobalContext} : list go.field_decl := [

--- a/types.go
+++ b/types.go
@@ -200,11 +200,6 @@ func (ctx *Ctx) namedRocqTypeDecl(spec *ast.TypeSpec) (decls []glang.Decl) {
 			fmt.Fprint(w, "|}.")
 
 			fmt.Fprint(w, "\n#[global] Arguments mk : clear implicits.")
-			fmt.Fprint(w, "\n#[global] Arguments t : clear implicits.")
-
-			fmt.Fprintf(w, "\nEnd def.\n")
-
-			fmt.Fprintf(w, "\nEnd %s.", glang.ToIdent(spec.Name.Name))
 		default:
 			fmt.Fprintf(w, "Definition t")
 
@@ -219,9 +214,10 @@ func (ctx *Ctx) namedRocqTypeDecl(spec *ast.TypeSpec) (decls []glang.Decl) {
 
 			rocqType := ctx.toGallinaType(spec, t)
 			fmt.Fprintf(w, "%s.", rocqType)
-			fmt.Fprint(w, "\nEnd def.")
-			fmt.Fprintf(w, "\nEnd %s.", glang.ToIdent(spec.Name.Name))
 		}
+		fmt.Fprint(w, "\n#[global] Arguments t : clear implicits.")
+		fmt.Fprint(w, "\nEnd def.")
+		fmt.Fprintf(w, "\nEnd %s.", glang.ToIdent(spec.Name.Name))
 	}
 
 	recordDecl := glang.VerbatimDecl{

